nupic/research/frameworks/vernon/mixins/step_based_logging.py

```diff
@@ -0,0 +1,143 @@
+]
+
+
+class StepBasedLogging(
+    interfaces.Experiment,  # Requires
+    interfaces.StepBasedLogging,  # Implements
+):
+    @staticmethod
+    def step_based_logging_interface_implemented():
+        return True
+
+    def setup_experiment(self, config):
+        """
+        :param config: Dictionary containing the configuration parameters
+
+            - log_timestep_freq: Configures mixins and subclasses that log every
+                                 timestep to only log every nth timestep (in
+                                 addition to the final timestep of each epoch).
+                                 Set to 0 to log only at the end of each epoch.
+        """
+        super().setup_experiment(config)
+        self._current_timestep = 0
+        self.log_timestep_freq = config.get("log_timestep_freq", 1)
+
+    @property
+    def current_timestep(self):
+        return self._current_timestep
+
+    @current_timestep.setter
+    def current_timestep(self, value):
+        self._current_timestep = value
+
+    def run_iteration(self):
+        timestep_begin = self.current_timestep
+        ret = super().run_iteration()
+        ret.update(
+            timestep_begin=timestep_begin,
+            timestep_end=self.current_timestep,
+        )
+        return ret
+
+    def post_batch(self, **kwargs):
+        super().post_batch(**kwargs)
+        # FIXME: move to post_optimizer_step
+        self.current_timestep += 1
+
+    def should_log_batch(self, train_batch_idx):
+        return (train_batch_idx == self.total_batches - 1) or (
+            self.log_timestep_freq > 0
+            and (self.current_timestep % self.log_timestep_freq) == 0)
+
+    def get_state(self):
+        state = super().get_state()
+        state["current_timestep"] = self.current_timestep
+        return state
+
+    def set_state(self, state):
+        super().set_state(state)
+        if "current_timestep" in state:
+            self.current_timestep = state["current_timestep"]
+
+    @classmethod
+    def get_recorded_timesteps(cls, result, config):
+        log_timestep_freq = config.get("log_timestep_freq", 1)
+        timestep_end = result["timestep_end"]
+        if log_timestep_freq == 0:
+            ret = [timestep_end - 1]
+        else:
+            # Find first logged timestep in range
+            logged_begin = int(math.ceil(result["timestep_begin"]
+                                         / log_timestep_freq)
+                               * log_timestep_freq)
+
+            ret = list(range(logged_begin, timestep_end, log_timestep_freq))
+
+            last_batch_timestep = timestep_end - 1
+            if last_batch_timestep % log_timestep_freq != 0:
+                ret.append(last_batch_timestep)
+
+        return ret
+
+    @classmethod
+    def expand_result_to_time_series(cls, result, config):
+        result_by_timestep = defaultdict(dict)
+
+        # Assign the epoch result to the appropriate timestep.
+        k_mapping = {
```

Comment:
Can we add a TODO or even a mental note for later to be able to define this `k_mapping` via the config or a classmethod?